p4-16/psa/PSA.mdk

```diff
@@ -523,34 +523,71 @@ extern Checksum<W> {
 
 ## Counters
 
-Counters are a simple mechanism for keeping statistics about the
-packets that trigger a table in a Match Action unit.
+Counters are a mechanism for keeping statistics.  The control plane
+can read counter values.  A P4 program cannot read counter values,
+only update them.  If you wish to implement a feature involving
+sequence numbers, for example, use Registers instead (Section
+[#sec-registers]).
+
+Direct counters are counters associated with a particular P4 table,
+and are implemented by the extern `DirectCounter`.  There are also
+'indirect' counters, which are implemented by the extern `Counter`.
+The primary differences between direct counters and indirect counters
+are:
+
+- Number of independently updatable counter values:
+  - A single instantiation of a direct counter always contains as many
+    independent counter values as the number of entries in the table
+    with which it is associated.
+  - You must specify the number of independent counter values for an
+    indirect counter when instantiating it.  This number of counters
+    need not be the same as the size of any P4 table.
+- Where counter updates are allowed in the P4 program:
+  - For a direct counter, you may only invoke its `count` method from
+    inside the actions of the table with which it is associated, and
+    this always updates the counter value associated with the matching
+    table entry.
+  - For an indirect counter, you may invoke its `count` method
+    anywhere in the P4 program where extern object method invocations
+    are permitted (e.g. inside actions, or directly inside a control's
+    `apply` block), and every such invocation must specify the index
+    of the counter value to be updated.
+
+One may implement packet counters by always using the `count` method
+that adds 1 to the counter value.
+
+One may implement byte counters by using the `count` method that adds
+a caller-specified `add_value` to the counter, containing a value that
+you wish to use for the length of the packet.
+
+We mention these as common use cases for counters.  You may of course
+use them to count things other than packets or bytes.
 
-Direct counters fire when the count method is invoked in an action, and
-have an instance for each entry in the table.
-
-
-### Counter types
-
-```
-enum CounterType_t {
-    packets,
-    bytes,
-    packets_and_bytes
-}
-```
 
 ### Counter
 
 ```
-extern Counter<W, S> {
```

Comment:
I am happy to follow whatever naming convention is decided upon.  I'd recommend a separate issue for it, since it will have a global effect on the document.